### PR TITLE
Change AgentProcessor logic to fix memory leak

### PR DESCRIPTION
**Pull request by ervteng**
In a recent change, the Agent ID now changes after every episode reset. This caused the dictionary of keys to grow inside the AgentProcessor. This PR cleans up the adding/deletion of items on reset boundaries, so that the dict doesn't grow.

### CODE REVIEW

File: ml-agents/mlagents/trainers/agent_processor.py [hunk 1]
```diff
@@ -159,11 +157,21 @@ def add_experiences(
                 elif not curr_agent_step.done:
                     self.episode_steps[global_id] += 1
 
-            self.last_step_result[global_id] = batched_step_result
+            for _gid in action_global_agent_ids:
+                if _gid in self.last_step_result:
+                    if not self.last_step_result[_gid][0].done:
+                        if "action" in take_action_outputs:
+                            self.policy.save_previous_action(
+                                [global_id], take_action_outputs["action"]
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/agent_processor.py]: _gid ?

File: ml-agents/mlagents/trainers/agent_processor.py [hunk 2]
```diff
@@ -159,11 +157,21 @@ def add_experiences(
...
+                            )
+                    else:
+                        # If it was done, delete it
+                        del self.last_step_result[_gid]
 
-        if "action" in take_action_outputs:
-            self.policy.save_previous_action(
-                previous_action.agent_ids, take_action_outputs["action"]
+            # Index is needed to grab from last_take_action_outputs
+            self.last_step_result[global_id] = (
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/agent_processor.py]: _gid ?

File: ml-agents/mlagents/trainers/agent_processor.py (hunk 1 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/agent_processor.py]: Yes

File: ml-agents/mlagents/trainers/agent_processor.py (hunk 2 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/agent_processor.py]: This one needs to be global_id, its in a different loop

**Comment by chriselion** [comment]
How do you know it works without a test?

**Comment by ervteng** [comment]
> How do you know it works without a test?

Tracking the size of the dicts using example envs (Hallway, Bouncer). Still need to think of a good unit test for this; we need to emulate On-demand decisions

**Comment by ervteng** [comment]
image
Bouncer
image
Hallway

**Comment by ervteng** [comment]
@chriselion 

> How do you know it works without a test?

Added a unit test that checks the basic deleting mechanism between resets.

**Comment by chriselion** [comment]
Great, just wanted to make sure we had some checks on it.

**Pull request merged**